[Sensor] upgrade draw_observation in python Sensor class

bigbike opened the pull request:
## Motivation and Context
This diff is trying to address the following issues:
- **Address the code duplication.** The old `draw_observation` in `Sensor` (python) has the same code logic as the `drawObservation()` in the `CameraSensor` class (C++). It is a type of code duplication.
- **Such code logic does not work for the FisheyeSensor introduced recently.** It works ONLY for the "Pinhole" or "Ortho" Camera.
- **The old dependency is wrong.**  Before the actual rendering logic was implemented in the `Renderer` class (C++). And `Sensor` class (both in C++ and python) depended on it. The correct dependency should be: Each **sensor type** implements the actually rendering logic (because different sensors (Pinhole, Fisheye etc.) have different, specific logic), and `Renderer` depends on the sensors. 
- **Upgrade Fisheye sensor and Camera sensor accordingly.**

<!--- Why is this change required? What problem does it solve? -->
<!--- Please link to an existing issue here if one exists. -->
<!--- (we recommend to have an existing issue for each pull request) -->

## How Has This Been Tested

<!--- Please describe here how your modifications have been tested. -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

bigbike commented on src/esp/gfx/Renderer.h:
```diff
@@ -38,11 +40,22 @@ class Renderer {
+  // TODO: deprecate the following function: draw(visualSensor, sceneGraph,
+  // flags);
   // draw the scene graph with the visual sensor provided by user
   void draw(sensor::VisualSensor& visualSensor,
```
My opinion: This function is really obsolete and thus should be removed.

bigbike commented:
Tested it with the `example.py` and it gave the correct result. See the attached picture:
![test rgba 00000](https://user-images.githubusercontent.com/931093/113498656-c637a080-94c3-11eb-8098-ecd7fe2b9ac1.png)

Skylion007 commented on habitat_sim/simulator.py:
```diff
@@ -553,45 +553,7 @@ def draw_observation(self) -> None:
-            and self._sim.get_active_scene_graph()
-            is not self._sim.get_active_semantic_scene_graph()
-        ):
-            agent_node.parent = self._sim.get_active_scene_graph().get_root_node()
```
Doesn't this need to be replicated somewhere in the C++? We have a couple of the tests are segfault with the new C++ function.

bigbike replied:
@Skylion007 : Yeah, this is the root cause I found last night.  See my previous comments.
Now I believe it is resolved in the `CameraSensor` class. Let us see what the CI would say.

bigbike commented on src/esp/sensor/CameraSensor.cpp:
```diff
@@ -115,20 +124,42 @@ bool CameraSensor::drawObservation(sim::Simulator& sim) {
+    bool twoSceneGraphs =
+        (&sim.getActiveSemanticSceneGraph() != &sim.getActiveSceneGraph());
+
+    if (twoSceneGraphs) {
+      // *assume* the sensor is always in the regular scene graph.
+      // backup first
+      Mn::Matrix4 relativeTransform = node().transformation();
+      Mn::Matrix4 absTransform = node().absoluteTransformation();
+      scene::SceneNode* p = static_cast<scene::SceneNode*>(node().parent());
+      // take the sensor from the current regular scene graph and connect it to
+      // the root node of semantic scene graph, set the *correct* transformation
+      node().setParent(&sim.getActiveSemanticSceneGraph().getRootNode());
+      node().setTransformation(absTransform);
+      draw(sim.getActiveSemanticSceneGraph(), flags);
+      // after drawing, put the node back to the regular scene graph.
+      // MUST use the setParent function! Otherwise the sensorSuite in the node
+      // will be wrong!!!
+      node().setParent(p);
+      node().setTransformation(relativeTransform);
+    } else {
+      draw(sim.getActiveSemanticSceneGraph(), flags);
+    }
```
@Skylion007 : I just updated the code as I replicated the logic here, and did it in the correct way with `setParent()`.

Skylion007 commented on src/esp/sensor/CameraSensor.cpp:
```diff
@@ -115,20 +124,42 @@ bool CameraSensor::drawObservation(sim::Simulator& sim) {
+    }
+
+    if (twoSceneGraphs) {
       flags |= gfx::RenderCamera::Flag::ObjectsOnly;
```
BTW, this cludge we are doing is going to hurt our FPS a lot in scenes without the semantic mesh, but only added objects (such as synthetic datasets.) One thing on the agenda is to find a better way to resolve this long term so we don't have to do this double pass all the time. @aclegg3 I think you mentioned something about this during our last meeting?

Skylion007 replied:
This can of course be saved for a different PR, just wanted to bring it up.

aclegg3 replied:
🤔  I'm not sure that this is the case. Here if we only have a single scene graph we'll just render it with semantic ids, if we have 2 scene-graphs (e.g. a specific separate semantic asset) then we render that asset and then a pass over dynamic objects. I don't see the overhead you are worried about. Am I missing something?

bigbike replied:
@aclegg3 : It means the absolute transformation for all nodes will be computed twice (one per each SG).

Skylion007 commented on habitat_sim/simulator.py:
```diff
@@ -553,45 +553,7 @@ def draw_observation(self) -> None:
 
-        # get the correct scene graph based on application
-        if self._spec.sensor_type == SensorType.SEMANTIC:
-            if self._sim.semantic_scene is None:
```
Do we need to add an ESPCHECK for this in the C++ or are we good?

bigbike replied:
@Skylion007 : ESP_CHECK is only used in the C++, NO?

Skylion007 replied:
Yeah, but this check was removed. I just want to make sure we throw the error somewhere.

bigbike replied:
OK, good catch. I will see if I can do something in the `getActiveSemanticSceneGraph`.

bigbike commented on src/esp/bindings/GfxBindings.cpp:
```diff
@@ -83,20 +84,26 @@ void initGfxBindings(py::module& m) {
           [](Renderer& self, sensor::VisualSensor& visualSensor,
-             scene::SceneGraph& sceneGraph, RenderCamera::Flag flags) {
-            self.draw(visualSensor, sceneGraph, RenderCamera::Flags{flags});
+             scene::SceneGraph& sceneGraph, RenderCamera::Flags flags) {
```
@aclegg3 : 
Would you mind taking a close look at it? 
The original code was introduced in #679 . I believe it is a bug. Can you help me to double verify it?
Thanks.

bigbike commented on src/esp/core/ManagedContainer.h:
```diff
@@ -703,7 +703,7 @@ class ManagedContainer : public ManagedContainerBase {
 
  public:
-  ESP_SMART_POINTERS(ManagedContainer<T, Access>);
+  ESP_SMART_POINTERS(ManagedContainer<T, Access>)
```
this ";" gives us a lot of compiling warnings. Get rid of it in this PR as well.

bigbike commented on src/esp/gfx/Renderer.cpp:
```diff
@@ -51,6 +53,16 @@ struct Renderer::Impl {
+  void draw(sensor::VisualSensor& visualSensor, sim::Simulator& sim) {
+    if (visualSensor.specification()->sensorType ==
+        sensor::SensorType::Semantic) {
+      ESP_CHECK(sim.semanticSceneExists(),
```
@Skylion007 : I added a sanity check here to address your concern.

Skylion007 commented on src/esp/sensor/FisheyeSensor.cpp:
```diff
@@ -155,12 +155,36 @@ bool FisheyeSensor::drawObservation(sim::Simulator& sim) {
+    if (twoSceneGraphs) {
+      // *assume* the sensor is always in the regular scene graph.
+      // backup first
+      Mn::Matrix4 relativeTransform = node().transformation();
```
Perhaps this should be abstracted out if possible. Disappointing that this hack is necessary, but it is required for every VisualSensor, right?

Skylion007 replied:
That can also be saved for a later PR though.

bigbike replied:
Yes, a bit code duplication here. I did not figure out a good way to extract it so far. Like I said in the description, different visual sensors has it is own rendering logic. Read this two pieces of code. They are a bit different.

bigbike replied:
Allow me to think about it.
This is problem that we need to solve.

Otherwise it will become a big problem as more sensors are coming in the future.

@mosra @aclegg3 @erikwijmans any ideas here?

bigbike replied:
I am thinking the ways to address this. It must be solved before committing this diff in.

bigbike commented on src/esp/sensor/FisheyeSensor.cpp:
```diff
@@ -147,12 +147,41 @@ bool FisheyeSensor::drawObservation(sim::Simulator& sim) {
+      moveSemanticSensorToSemanticSceneGraph(sim);
+      cubeMap_->renderToTexture(*cubeMapCamera_,
+                                sim.getActiveSemanticSceneGraph(), flags);
+      moveSemanticSensorBackToRegularSceneGraph(sim);
```
@Skylion007 : I reduced the code duplication by extracting 2 helper functions applied here.

The challenge is that the code looks like a "sandwich". The "bread" is the same but the "food" in between varies.

Skylion007 commented:
Similar to render_enter / exit semantics  though, right?  It would need to
be split between two functions. Also could be done with a functional
programming lambda interface, although not as cleanly.

On Mon, Apr 5, 2021 at 9:58 PM bigbike ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/esp/sensor/FisheyeSensor.cpp
> <https://github.com/facebookresearch/habitat-sim/pull/1159#discussion_r607435548>
> :
>
> >    // generate the cubemap texture
> -  cubeMap_->renderToTexture(*cubeMapCamera_, sim.getActiveSceneGraph(), flags);
> +  if (fisheyeSensorSpec_->sensorType == SensorType::Semantic) {
> +    bool twoSceneGraphs =
> +        (&sim.getActiveSemanticSceneGraph() != &sim.getActiveSceneGraph());
> +
> +    if (twoSceneGraphs) {
> +      moveSemanticSensorToSemanticSceneGraph(sim);
> +      cubeMap_->renderToTexture(*cubeMapCamera_,
> +                                sim.getActiveSemanticSceneGraph(), flags);
> +      moveSemanticSensorBackToRegularSceneGraph(sim);
>
> @Skylion007 <https://github.com/Skylion007> : I reduced the code
> duplication by extracting 2 helper functions applied here.
>
> The challenge is that the code looks like a "sandwich". The "bread" is the
> same but the "food" in between varies.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebookresearch/habitat-sim/pull/1159#pullrequestreview-628453497>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPVMXYTQYRXR2OM7GVJHULTHJTEDANCNFSM42K66TIA>
> .
>

erikwijmans commented on src/esp/sensor/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ add_library(
 target_link_libraries(
   sensor
   PUBLIC core gfx scene
+  PRIVATE sim
```
Why private?

bigbike replied:
It was set (in different ways) to help me better understand why the CI fails.
It is a sad story. I was taking PTO on Mon. but I spent whole day trying to solve the CI failure, which is still there.

I found the `ResourceMangerTest` cannot be compiled. It complains `libsensor.a` is missing some definitions from here and there, even though it does not need anything from `sensor`.

erikwijmans commented on src/esp/sensor/CameraSensor.cpp:
```diff
@@ -107,6 +106,16 @@ gfx::RenderCamera* CameraSensor::getRenderCamera() const {
+void CameraSensor::draw(scene::SceneGraph& sceneGraph,
+                        gfx::RenderCamera::Flags flags) {
+  for (auto& it : sceneGraph.getDrawableGroups()) {
+    // TODO: remove || true
```
This TODO is getting carried around more and more places, are we ever going to actually do it?

bigbike replied:
I believe the purpose is to build a filter function on drawables, which can be useful in the material editing.

erikwijmans replied:
Yeah, the `it.second.prepareForDraw(*renderCamera_)` part of the if statement seems fine and I can see a use for that returning false and something not getting drawn, my concern is that we are simply ignoring the return value of that function via the `|| true`.  If we ever do want to do something with that function, we have a bunch of ` || true`'s that will make that harder.  

I don't know the history of the `|| true` (I think it's been there forever and thus likely a debugging thing that never got removed), but it currently does nothing since the only implementation for `prepareForDraw` just returns true.

erikwijmans commented on src/esp/sim/Simulator.h:
```diff
@@ -73,6 +73,8 @@ class Simulator {
     return semanticScene_;
   }
 
+  bool semanticSceneExists() { return (semanticScene_ != nullptr); }
```
```suggestion
  bool semanticSceneExists() const { return (semanticScene_ != nullptr); }
```

erikwijmans commented on src/esp/sensor/VisualSensor.cpp:
```diff
@@ -136,5 +137,39 @@ bool VisualSensor::getObservation(sim::Simulator& sim, Observation& obs) {
+    return;
+  }
+
+  node().setParent(semanticSensorParentNodeBackup_);
```
Can you add an assert that the pointer is not null and the optional is not empty and then null/empty them after?

bigbike replied:
OK. Will do.

erikwijmans commented on src/esp/sensor/CameraSensor.cpp:
```diff
@@ -115,20 +124,30 @@ bool CameraSensor::drawObservation(sim::Simulator& sim) {
+        (&sim.getActiveSemanticSceneGraph() != &sim.getActiveSceneGraph());
+
+    if (twoSceneGraphs) {
+      moveSemanticSensorToSemanticSceneGraph(sim);
```
This seems a bit risky as it's easy to forget to call the `moveSemanticSensorBackToRegularSceneGraph` function.  Worth considering RAII here.

bigbike replied:
Can you elaborate a bit how you are going to do RAII here?

if people forgets to call this function `moveSemanticSensorBackToRegularSceneGraph`, she will know immediately as in the following (L145), she needs it to draw the "objects". If the above function is not called, the program will hit an assertion in Magnum (camera is not in the same SG) and quit.

erikwijmans replied:
Just a helper class that calls `moveSemanticSensorToSemanticSceneGraph` in its constructor and `moveSemanticSensorBackToRegularSceneGraph` in its destructor.

Skylion007 commented:
@bigbike Merging from master should solve the CI issues caused by doc building

Skylion007 commented:
Okay weird, only 1 test is running.

erikwijmans commented on src/esp/sensor/VisualSensor.cpp:
```diff
@@ -136,5 +137,58 @@ bool VisualSensor::getObservation(sim::Simulator& sim, Observation& obs) {
+      "sensor is attached to the root node, and thus cannot be moved.", );
+
+  // check if the sensor is already in this semantic scene graph
+  if (node.scene() == sim.getActiveSemanticSceneGraph().getRootNode().scene()) {
```
I think this case will always cause a crash?  Since the destructor will always hit `CORRADE_INTERNAL_ASSERT(semanticSensorParentNodeBackup_);`

bigbike replied:
Well, the basic assumption is that at the very beginning, the semantic sensor is attached to the regular scene graph, not the semantic scene graph. Your comment makes me think if it is always true.

If this is true, then you do not need to worry about it here.

bigbike replied:
If there are NO two scene graphs, e.g., only the semantic scene graph, then user should never ever use `MoveSemanticSensorNodeHelper` at all.

erikwijmans replied:
Should this be an assert then?  Right now it looks like a feature where this won't move the sensor if it doesn't need to.  However, the destructor assumes that if the sensor isn't part of the active scene graph, then the constructor moved it to the semantic scene graph and will try to move it back.  This will cause a crash however.

erikwijmans commented on src/esp/sensor/VisualSensor.h:
```diff
@@ -195,6 +199,34 @@ class VisualSensor : public Sensor {
   VisualSensorSpec::ptr visualSensorSpec_ =
       std::dynamic_pointer_cast<VisualSensorSpec>(spec_);
+
+  class MoveSemanticSensorNodeHelper {
```
Yeah, this looks nice.  Left one question about implementation

bigbike commented on src/esp/sensor/CameraSensor.cpp:
```diff
@@ -115,20 +123,29 @@ bool CameraSensor::drawObservation(sim::Simulator& sim) {
-    // SensorType is Depth or any other type
-    renderer->draw(*this, sim.getActiveSceneGraph(), flags);
+    // SensorType is Color, Depth or any other type
+    draw(sim.getActiveSceneGraph(), flags);
```
@erikwijmans :
If there is only one scene graph, we assume it is the regular scene graph.
Is there a chance that we will only have semantic scene graph?
also cc: @aclegg3

aclegg3 replied:
If we only have a one scene graph we'll use it for both RGB and SEMANTIC and all object will be on the correct scene graph for both renders.

bigbike commented:
Dear reviewers,

If you do not have more concerns, I am going to merge it in this afternoon. Thanks.
@erikwijmans : hope the latest version addressed all your concerns.

mosra reviewed (approved):
Just a random comment that doesn't affect the approval status.

erikwijmans reviewed (approved):
🚢

Skylion007 commented:
Looks good. Let's merge this in.
